Gas Optimisations.md

```diff
@@ -18,36 +18,56 @@ Recall that structs and arrays start at a diff slot. So inside a struct declarin
 
 But, `uint128, uint256, uint128` is worst due to **3 slots** as storage make a **new slot** if a variable can't fit.
 
-## Creating a function for encoding function selector : 
+## Avoid calculating function selectors at runtime:
 
-Instead of `addr.call(abi.encodeWithSignature("transfer(address,uint256)", 0xSomeAddress, 123))`, we can prevent computing the signature everytime and let a function do it like this,
+The compiler can precalculate a function selector for at compilation time you when the function signature is known.
+Try to rely on this behavior whenever possible.
 
+For example the following code prevents this optimization:
 ```solidity
-function getSelector(string calldata _func) external pure returns (bytes4) {
-        return bytes4(keccak256(bytes(_func)));
-    }
+function getSelector(bytes memory _func) returns (bytes4) {
+    return bytes4(keccak256(_func));
+}
+
+function doStuff() {
+    addr.call(abi.encodeWithSelector(getSelector("transfer(address,uint256)"), 0xSomeAddress, 123));
+}
 ```
+`getSelector()` will always run `keccak256()` on its argument because it cannot assume that its input is always a constant.
+With `abi.encodeWithSignature()`, on the other hand, the compiler is smart enough to generate more efficient code when you use a string literal.
 
-And use `addr.call(getSelector(funcName), 0xSomeAddress, 123))`. This saves a tiny amount of gas.
+```selector
+addr.call(abi.encodeWithSignature("transfer(address,uint256)", 0xSomeAddress, 123));
+```
 
+If you inspect the resulting assembly, you'll notice that the value `0x6628616464726573732c75696e7429` (i.e. `"transfer(address,uint256)"') is nowhere to be found and the result of the `keccak256()` call (`0xa9059cbb`) is hard-coded in it instead.
 
+**Note**: Calculating selectors by hand is very error-prone and should be avoided.
+A very common mistake is to insert a space after the comma, which will result in a completely different selector.
+The example above does such a calculation on purpose, to illustrate a point, but in practice the language provides you with a more type-safe way to achieve the same result: the `.selector` member:
 
-- **Constants** don't occupy any storage. They're always **replaced when needed in the code like a macro**. 
+```selector
+addr.call(abi.encodeWithSelector(IERC20.transfer.selector, 0xSomeAddress, 123));
+```
```

Comment:
This part in particular might require some discussion because I might have misinterpreted what you meant. For example your `getSelector()` being `external` indicates that it's meant to be invoked by another contract and I don't see why anyone would want that, which makes me think I'm missing something here.

But anyway, as I understand it, you're recommending a helper for calculating the selector and this will actually have the opposite result - it will prevent the compiler from performing an optimization here because it can no longer assume that the signature is a compile-time constant.

I have rewritten it to instead point out that selectors should not be hashed at runtime, but I'm not sure that this recommendation is overly useful. Normally you would use `abi.encodeWithSelector()` combined with the `.selector` rather than `abi.encodeWithSignature()` and you would not have to deal with signatures at all - unless you actually receive the signature as input but they it's no longer a constant anyway.

It's also worth noting that Solidity 0.8.11 offers the new `abi.encodeCall()` helper (see [ABI Encoding and Decoding Functions](https://docs.soliditylang.org/en/latest/units-and-global-variables.html#abi-encoding-and-decoding-functions)) which is recommended for encoding the input to `call()`. I did not mention it in the text because unfortunately it has an unfortunate flaw: it requires a contract instance rather than a type. It's usable creating an instance (e.g. something like `abi.encodeCall(IERC20(addr).transfer, ...)` but it's not a canonical usage and I'd not put that in a general guide. This will be fixed in 0.8.12.